chain/network/src/routing.rs

```diff
@@ -758,100 +780,175 @@ pub struct RoutingTableInfo {
-
-        distance.insert(&self.source, 0);
-
-        // Add active connections
-        if let Some(neighbors) = self.adjacency.get(&self.source) {
-            for neighbor in neighbors {
-                queue.push(neighbor);
-                distance.insert(neighbor, 1);
-                routes.insert(neighbor.clone(), vec![neighbor.clone()].drain(..).collect());
+    pub fn calculate_distance(&self) -> HashMap<PeerId, Vec<PeerId>> {
+        // TODO add removal of unreachable nodes
+
+        let mut queue = VecDeque::new();
+
+        let nodes = self.unused.len() + self.id2p.len();
+        let mut distance: Vec<i32> = vec![-1; nodes];
+        let mut routes: Vec<u128> = vec![0; nodes];
+
+        let source_id = 0;
+        distance[source_id] = 0;
+
+        if let Some(neighbors) = self.adjacency.get(&source_id) {
+            for (id, neighbor) in neighbors.iter().enumerate().take(128) {
+                queue.push_back(*neighbor);
+                distance[*neighbor] = 1;
+                routes[*neighbor] = (1 as u128) << id;
             }
         }
 
-        let mut head = 0;
-
-        while head < queue.len() {
-            let cur_peer = queue[head];
-            let cur_distance = *distance.get(cur_peer).unwrap();
-            head += 1;
+        while let Some(cur_peer) = queue.pop_front() {
+            let cur_distance = distance[cur_peer];
 
             if let Some(neighbors) = self.adjacency.get(&cur_peer) {
                 for neighbor in neighbors {
-                    if let Entry::Vacant(entry) = distance.entry(neighbor) {
-                        queue.push(entry.key());
-                        entry.insert(cur_distance + 1);
-                        routes.insert(neighbor.clone(), HashSet::new());
+                    if distance[*neighbor] == -1 {
+                        distance[*neighbor] = cur_distance + 1;
+                        queue.push_back(*neighbor);
                     }
-
                     // If this edge belong to a shortest path, all paths to
                     // the closer nodes are also valid for the current node.
-                    if *distance.get(neighbor).unwrap() == cur_distance + 1 {
-                        let adding_routes = routes.get(cur_peer).unwrap().clone();
-                        let target_routes = routes.get_mut(neighbor).unwrap();
-
-                        for route in adding_routes {
-                            target_routes.insert(route.clone());
-                        }
+                    if distance[*neighbor] == cur_distance + 1 {
+                        routes[*neighbor] |= routes[cur_peer];
                     }
                 }
             }
         }
 
-        routes.into_iter().filter(|(_, hops)| !hops.is_empty()).collect()
+        self.compute_result(&mut routes, &distance)
+    }
+
+    fn compute_result(
+        &self,
+        routes: &mut Vec<u128>,
+        distance: &Vec<i32>,
```

Comment:
```suggestion
        routes: &[u128],
        distance: &[i32],
```

`&Vec<_>` is a "useless type": https://github.com/rust-analyzer/rust-analyzer/blob/master/docs/dev/style.md#useless-types